// ng2-materialize/dist/card/card.component.js
import { Component, Directive, Input, ViewChild } from '@angular/core';
var MzCardComponent = (function () {
    function MzCardComponent() {
    }
    MzCardComponent.prototype.ngAfterViewInit = function () {
        this.initCardAction();
    };
    MzCardComponent.prototype.initCardAction = function () {
        var cardActionElement = $(this.cardActionWrapper.nativeElement).find('mz-card-action');
        // hide action wrapper when mz-card-action element is not present
        // or when mz-card-action does not contain any html (text or elements)
        if (cardActionElement.length === 0 || cardActionElement[0].innerHTML.trim() === '') {
            this.cardActionWrapper.nativeElement.style.display = 'none';
        }
    };
    return MzCardComponent;
}());
export { MzCardComponent };
MzCardComponent.decorators = [
    { type: Component, args: [{
                selector: 'mz-card',
                template: "<div class=\"card {{ backgroundClass }}\" [class.hoverable]=\"hoverable\"><div class=\"card-content {{ textClass }}\"><span class=\"card-title\"><ng-content select=\"mz-card-title\"></ng-content></span><p><ng-content select=\"mz-card-content\"></ng-content></p></div><div #cardActionWrapper class=\"card-action\"><ng-content select=\"mz-card-action\"></ng-content></div></div>",
                styles: [""],
            },] },
];
/** @nocollapse */
MzCardComponent.ctorParameters = function () { return []; };
MzCardComponent.propDecorators = {
    'backgroundClass': [{ type: Input },],
    'hoverable': [{ type: Input },],
    'textClass': [{ type: Input },],
    'cardActionWrapper': [{ type: ViewChild, args: ['cardActionWrapper',] },],
};
// Declare the tags to avoid error: '<mz-card-x>' is not a known element
// https://github.com/angular/angular/issues/11251
// tslint:disable: directive-selector
var MzCardTitleDirective = (function () {
    function MzCardTitleDirective() {
    }
    return MzCardTitleDirective;
}());
export { MzCardTitleDirective };
MzCardTitleDirective.decorators = [
    { type: Directive, args: [{ selector: 'mz-card-title' },] },
];
/** @nocollapse */
MzCardTitleDirective.ctorParameters = function () { return []; };
var MzCardContentDirective = (function () {
    function MzCardContentDirective() {
    }
    return MzCardContentDirective;
}());
export { MzCardContentDirective };
MzCardContentDirective.decorators = [
    { type: Directive, args: [{ selector: 'mz-card-content' },] },
];
/** @nocollapse */
MzCardContentDirective.ctorParameters = function () { return []; };
var MzCardActionDirective = (function () {
    function MzCardActionDirective() {
    }
    return MzCardActionDirective;
}());
export { MzCardActionDirective };
MzCardActionDirective.decorators = [
    { type: Directive, args: [{ selector: 'mz-card-action' },] },
];
/** @nocollapse */
MzCardActionDirective.ctorParameters = function () { return []; };
